[Bugs] TestReSynchronization is failed with go1.23.0

## Describe the bug
The test TestReSynchronization is failed with go1.23.0, but is passed with go1.22.6.

## To Reproduce
1. Install go 1.23.0
2. Build UPF
3. Run command `./test.sh TestReSynchronization`

## Environment (please complete the following information):
 - free5GC Version: v3.4.2
 - OS: Ubuntu 20.04.6 LTS
 - Kernel version: 5.15.0-117-generic
 - go version: go version go1.23.0 linux/amd64

### Log File
[test.log](https://github.com/user-attachments/files/16656072/test.log)

We insert debug log to test function.
Strangely we observe crash with go 1.21.13.

```
$
$
$ git diff
diff --git a/test/registration_test.go b/test/registration_test.go
index 761e064..ba26800 100644
--- a/test/registration_test.go
+++ b/test/registration_test.go
@@ -2605,6 +2605,7 @@ func TestReSynchronization(t *testing.T) {
        }

        newSqnMs, _ := hex.DecodeString(newSqnMsString)
+       fmt.Print(newSqnMs)
        MAC_A, MAC_S := make([]byte, 8), make([]byte, 8)
        CK, IK := make([]byte, 16), make([]byte, 16)
        RES := make([]byte, 8)
$ go version
go version go1.21.13 linux/amd64
$ ./test.sh TestReSynchronization
2024-08-21T10:09:15.723231216+09:00 [INFO][UPF][Main] UPF version:
        free5GC version: v3.4.1
        build time:      2024-08-21T00:19:20Z
        commit hash:     705e50db
        commit time:     2024-03-27T16:48:24Z
        go version:      go1.21.13 linux/amd64
2024-08-21T10:09:15.723441122+09:00 [INFO][UPF][CFG] Read config from [./config/upfcfg.test.yaml]
2024-08-21T10:09:15.724063797+09:00 [INFO][UPF][CFG] ==================================================
2024-08-21T10:09:15.724126735+09:00 [INFO][UPF][CFG] (*factory.Config)(0xc00053a320)({
        Version: (string) (len=5) "1.0.3",
        Description: (string) (len=31) "UPF initial local configuration",
        Pfcp: (*factory.Pfcp)(0xc000527590)({
                Addr: (string) (len=14) "10.200.200.101",
                NodeID: (string) (len=14) "10.200.200.101",
                RetransTimeout: (time.Duration) 1s,
                MaxRetrans: (uint8) 3
        }),
        Gtpu: (*factory.Gtpu)(0xc000527740)({
                Forwarder: (string) (len=5) "gtp5g",
                IfList: ([]factory.IfInfo) (len=1 cap=1) {
                        (factory.IfInfo) {
                                Addr: (string) (len=14) "10.200.200.102",
                                Type: (string) (len=2) "N3",
                                Name: (string) "",
                                IfName: (string) "",
                                MTU: (uint32) 0
                        }
                }
        }),
        DnnList: ([]factory.DnnList) (len=1 cap=1) {
                (factory.DnnList) {
                        Dnn: (string) (len=8) "internet",
                        Cidr: (string) (len=12) "10.60.0.0/16",
                        NatIfName: (string) ""
                }
        },
        Logger: (*factory.Logger)(0xc000025620)({
                Enable: (bool) true,
                Level: (string) (len=4) "info",
                ReportCaller: (bool) false
        })
})
2024-08-21T10:09:15.724350055+09:00 [INFO][UPF][CFG] ==================================================
2024-08-21T10:09:15.724376516+09:00 [INFO][UPF][Main] Log level is set to [info]
2024-08-21T10:09:15.724399319+09:00 [INFO][UPF][Main] Report Caller is set to [false]
2024-08-21T10:09:15.724462584+09:00 [INFO][UPF][Main] starting Gtpu Forwarder [gtp5g]
2024-08-21T10:09:15.724500547+09:00 [INFO][UPF][Main] GTP Address: "10.200.200.102:2152"
2024-08-21T10:09:15.740441282+09:00 [INFO][UPF][BUFF] buff netlink server started
2024-08-21T10:09:15.740538379+09:00 [INFO][UPF][Perio] perio server started
2024-08-21T10:09:15.740568217+09:00 [INFO][UPF][Gtp5g] Forwarder started
2024-08-21T10:09:15.741041157+09:00 [INFO][UPF][PFCP][LAddr:10.200.200.101:8805] starting pfcp server
2024-08-21T10:09:15.741097475+09:00 [INFO][UPF][PFCP][LAddr:10.200.200.101:8805] pfcp server started
2024-08-21T10:09:15.741124591+09:00 [INFO][UPF][Main] UPF started
Run TestReSynchronization
2024-08-21T10:09:21.205653143+09:00 [INFO][NRF][Main] Log enable is set to [true]
2024-08-21T10:09:21.207002569+09:00 [INFO][NRF][Main] Log level is set to [info]
2024-08-21T10:09:21.207151835+09:00 [INFO][NRF][Main] Report Caller is set to [false]
2024-08-21T10:09:21.207299622+09:00 [INFO][NRF][Init] nrfconfig Info: Version[1.0.2] Description[NRF initial test configuration]
AMF Config failed: Invalid T3555: non zero value required
2024-08-21T10:09:21.209074716+09:00 [INFO][AMF][Main] Log enable is set to [true]
2024-08-21T10:09:21.209211731+09:00 [INFO][AMF][Main] Log level is set to [info]
2024-08-21T10:09:21.209343576+09:00 [INFO][AMF][Main] Report Caller is set to [false]
2024-08-21T10:09:21.209493681+09:00 [INFO][AMF][Util] amfconfig Info: Version[1.0.9]
2024-08-21T10:09:21.211524817+09:00 [INFO][SMF][Main] Log enable is set to [true]
2024-08-21T10:09:21.211710664+09:00 [INFO][SMF][Main] Log level is set to [info]
2024-08-21T10:09:21.211742881+09:00 [INFO][SMF][Main] Report Caller is set to [false]
2024-08-21T10:09:21.212305162+09:00 [INFO][UDR][Main] Log enable is set to [true]
2024-08-21T10:09:21.212479474+09:00 [INFO][UDR][Main] Log level is set to [info]
2024-08-21T10:09:21.212511579+09:00 [INFO][UDR][Main] Report Caller is set to [false]
2024-08-21T10:09:21.214551938+09:00 [INFO][PCF][Main] Log enable is set to [true]
2024-08-21T10:09:21.214594498+09:00 [INFO][PCF][Main] Log level is set to [info]
2024-08-21T10:09:21.214713328+09:00 [INFO][PCF][Main] Report Caller is set to [false]
2024-08-21T10:09:21.214819407+09:00 [INFO][PCF][Util] pcfconfig Info: Version[1.0.2] Description[PCF initial test configuration]
2024-08-21T10:09:21.216080253+09:00 [INFO][UDM][Main] Log enable is set to [true]
2024-08-21T10:09:21.216122033+09:00 [INFO][UDM][Main] Log level is set to [info]
2024-08-21T10:09:21.216227474+09:00 [INFO][UDM][Main] Report Caller is set to [false]
2024-08-21T10:09:21.216452207+09:00 [INFO][NSSF][Main] Log enable is set to [true]
2024-08-21T10:09:21.216480880+09:00 [INFO][NSSF][Main] Log level is set to [info]
2024-08-21T10:09:21.216573981+09:00 [INFO][NSSF][Main] Report Caller is set to [false]
2024-08-21T10:09:21.217226491+09:00 [INFO][AUSF][Main] Log enable is set to [true]
2024-08-21T10:09:21.217381080+09:00 [INFO][AUSF][Main] Log level is set to [info]
2024-08-21T10:09:21.217515740+09:00 [INFO][AUSF][Main] Report Caller is set to [false]
2024-08-21T10:09:21.217733050+09:00 [INFO][AUSF][Init] ausfconfig Info: Version[1.0.3] Description[AUSF initial test configuration]
ausf context =  &{{{0 0} {[] {} <nil>} map[] 0} {{0 0} {[] {} <nil>} map[] 0} 7a7aa04e-22d7-48e2-894f-63d48cc59b77 ausfGroup001 8000 127.0.0.9 127.0.0.9 http://127.0.0.9:8000 http http://127.0.0.10:8000 ../cert/nrf.pem map[nausf-auth:{7a7aa04e-22d7-48e2-894f-63d48cc59b77 nausf-auth 0xc00028a828 http REGISTERED   0xc00028a810  [] <nil> [] [] <nil> 0 0 0 <nil> <nil>  }] [{208 93} {123 45}]  0xc0000af4a0 false false}
2024-08-21T10:09:21.218957763+09:00 [INFO][CHF][Main] Log enable is set to [true]
2024-08-21T10:09:21.219132165+09:00 [INFO][CHF][Main] Log level is set to [info]
2024-08-21T10:09:21.219273107+09:00 [INFO][CHF][Main] Report Caller is set to [false]
2024-08-21T10:09:21.219414672+09:00 [INFO][CHF][Init] chfconfig Info: Version[1.0.3] Description[CHF initial test configuration]
chf context =  &{f92434e2-90f6-4296-98a0-f8c7e1709bc5 CHF http://127.0.0.113:8000 http 127.0.0.113 127.0.0.113 8000 map[nchf-convergedcharging:{f92434e2-90f6-4296-98a0-f8c7e1709bc5 nchf-convergedcharging 0xc00028bb00 http REGISTERED   0xc00028bae8 http://127.0.0.113:8000 [] <nil> [] [] <nil> 0 0 0 <nil> <nil>  }] map[] 0 http://127.0.0.10:8000 ../cert/nrf.pem {{0 0} {[] {} <nil>} map[] 0} false 0xc0006bce40 0xc0006bcea0 0xc0002b52c0 0xc0002b5380}
2024-08-21T10:09:21.219922888+09:00 [INFO][NRF][Init] Server starting
2024-08-21T10:09:21.220248594+09:00 [INFO][NRF][Init] Binding addr: [127.0.0.10:8000]
2024-08-21T10:09:21.420274311+09:00 [INFO][AMF][Init] Server started
2024-08-21T10:09:21.420557023+09:00 [INFO][AMF][Util] amfconfig Info: Version[1.0.9]
2024-08-21T10:09:21.421101579+09:00 [INFO][AMF][Ngap] Listen on 127.0.0.1:38412
2024-08-21T10:09:21.423270338+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:21.426859796+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:21.428129535+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:21.430402723+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/c42281b0-b8a8-4a39-9185-1f24d515ffb2
2024-08-21T10:09:21.430743048+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/c42281b0-b8a8-4a39-9185-1f24d515ffb2 |
2024-08-21T10:09:21.431380520+09:00 [INFO][AMF][Main] OAuth2 setting receive from NRF: false
2024-08-21T10:09:21.621229255+09:00 [INFO][SMF][CTX] smfconfig Info: Version[1.0.7] Description[SMF initial single test configuration]
2024-08-21T10:09:21.621621187+09:00 [INFO][SMF][CTX] Endpoints: [10.200.200.102]
2024-08-21T10:09:21.621864045+09:00 [INFO][SMF][Init] Server started
2024-08-21T10:09:21.623361090+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:21.625865273+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:21.626931145+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:21.628704894+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/099a96e5-5d1a-422f-8f8c-49226d996adc
2024-08-21T10:09:21.629197394+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/099a96e5-5d1a-422f-8f8c-49226d996adc |
2024-08-21T10:09:21.630002184+09:00 [INFO][SMF][Main] OAuth2 setting receive from NRF: false
2024-08-21T10:09:21.630076291+09:00 [INFO][SMF][Init] SMF Registration to NRF {099a96e5-5d1a-422f-8f8c-49226d996adc SMF REGISTERED 0 0xc0001797d0 0xc0001797e8 [] []   [127.0.0.2] [] <nil> [] [] <nil> 0 0 0 area1 <nil> <nil> <nil> <nil> 0xc000b2d900 <nil> <nil> <nil> <nil> <nil> map[oauth2:false] <nil> false 0xc000179728 false false []}
2024-08-21T10:09:21.630276557+09:00 [INFO][SMF][PFCP] Listen on 10.200.200.1:8805
2024-08-21T10:09:21.630326904+09:00 [INFO][SMF][Main] Sending PFCP Association Request to UPF[10.200.200.101]
2024-08-21T10:09:21.630764687+09:00 [INFO][UPF][PFCP][LAddr:10.200.200.101:8805] handleAssociationSetupRequest
2024-08-21T10:09:21.631039803+09:00 [INFO][UPF][PFCP][LAddr:10.200.200.101:8805][CPNodeID:10.200.200.1] New node
2024-08-21T10:09:21.631380109+09:00 [INFO][SMF][Main] Received PFCP Association Setup Accepted Response from UPF[10.200.200.101]
2024-08-21T10:09:21.822466115+09:00 [INFO][UDR][Init] UDR Config Info: Version[1.0.2] Description[UDR initial test configuration]
2024-08-21T10:09:21.822602756+09:00 [INFO][UDR][Init] Server started
2024-08-21T10:09:21.822994957+09:00 [INFO][UDR][Util] udrconfig Info: Version[1.0.2] Description[UDR initial test configuration]
2024-08-21T10:09:21.824079178+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:21.826252324+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:21.827223926+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:21.828566741+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/e5d91a11-c767-4e52-bd02-a3e1647e3775
2024-08-21T10:09:21.828692426+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/e5d91a11-c767-4e52-bd02-a3e1647e3775 |
2024-08-21T10:09:21.829064648+09:00 [INFO][UDR][Main] OAuth2 setting receive from NRF: false
2024-08-21T10:09:22.023168081+09:00 [INFO][PCF][Init] Server started
2024-08-21T10:09:22.023394843+09:00 [INFO][PCF][Util] pcfconfig Info: Version[1.0.2] Description[PCF initial test configuration]
2024-08-21T10:09:22.024738088+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:22.026872580+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:22.028097298+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:22.029108473+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/38e79c26-c3c5-4059-bbb5-06f02340ea08
2024-08-21T10:09:22.029257875+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/38e79c26-c3c5-4059-bbb5-06f02340ea08 |
2024-08-21T10:09:22.029748382+09:00 [INFO][PCF][Main] OAuth2 setting receive from NRF: false
2024-08-21T10:09:22.030118902+09:00 [INFO][NRF][DISC] Handle NFDiscoveryRequest
2024-08-21T10:09:22.031620397+09:00 [INFO][NRF][GIN] | 200 |       127.0.0.1 | GET     | /nnrf-disc/v1/nf-instances?requester-nf-type=PCF&service-names=nudr-dr&target-nf-type=UDR |
2024-08-21T10:09:22.224407749+09:00 [INFO][UDM][Init] UDM Config Info: Version[1.0.3] Description[UDM initial test configuration]
2024-08-21T10:09:22.224532190+09:00 [INFO][UDM][Init] Server started
2024-08-21T10:09:22.224687920+09:00 [INFO][UDM][Util] udmconfig Info: Version[1.0.3] Description[UDM initial test configuration]
2024-08-21T10:09:22.226199577+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:22.228314581+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:22.229479393+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:22.230960829+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/a79d969e-f104-4ecb-a93a-8dfb97aa21e2
2024-08-21T10:09:22.231161266+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/a79d969e-f104-4ecb-a93a-8dfb97aa21e2 |
2024-08-21T10:09:22.231685205+09:00 [INFO][UDM][Main] OAuth2 setting receive from NRF: false
2024-08-21T10:09:22.425795680+09:00 [INFO][NSSF][Init] Server started
2024-08-21T10:09:22.427133393+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:22.429202869+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:22.430265055+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:22.431555445+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/5da6a869-2a03-41b6-a325-d0b9a8db3b07
2024-08-21T10:09:22.431692767+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/5da6a869-2a03-41b6-a325-d0b9a8db3b07 |
2024-08-21T10:09:22.432061229+09:00 [INFO][NSSF][Main] OAuth2 setting receive from NRF: false
2024-08-21T10:09:22.626257672+09:00 [INFO][AUSF][Init] Server started
2024-08-21T10:09:22.627654569+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:22.630331875+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:22.631338897+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:22.632713998+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/7a7aa04e-22d7-48e2-894f-63d48cc59b77
2024-08-21T10:09:22.632825436+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/7a7aa04e-22d7-48e2-894f-63d48cc59b77 |
2024-08-21T10:09:22.633253449+09:00 [INFO][AUSF][Main] OAuth2 setting receive from NRF: false
2024-08-21T10:09:22.827335936+09:00 [INFO][CHF][Init] Server started
2024-08-21T10:09:22.828151798+09:00 [INFO][CHF][CGF] FTP server Start
2024-08-21T10:09:22.828483554+09:00 [WARN][CHF][CGF] Login to Webconsole FTP fail: dial tcp 127.0.0.1:2122: connect: connection refused, retrying [1]
2024-08-21T10:09:22.829775413+09:00 [INFO][CHF][Acct] Open Account Balance Management Server
2024-08-21T10:09:22.833591084+09:00 [INFO][NRF][NFM] Handle NFRegisterRequest
2024-08-21T10:09:22.835685926+09:00 [INFO][NRF][NFM] urilist update
2024-08-21T10:09:22.836736082+09:00 [INFO][NRF][NFM] Create NF Profile
2024-08-21T10:09:22.837974055+09:00 [INFO][NRF][NFM] Location header:  http://127.0.0.10:8000/nnrf-nfm/v1/nf-instances/f92434e2-90f6-4296-98a0-f8c7e1709bc5
2024-08-21T10:09:22.838150695+09:00 [INFO][NRF][GIN] | 201 |       127.0.0.1 | PUT     | /nnrf-nfm/v1/nf-instances/f92434e2-90f6-4296-98a0-f8c7e1709bc5 |
2024-08-21T10:09:22.838572841+09:00 [INFO][CHF][Main] OAuth2 setting receive from NRF: false
=== RUN   TestReSynchronization
2024-08-21T10:09:23.031773502+09:00 [INFO][AMF][Ngap] [AMF] SCTP Accept from: 127.0.0.1:9487
2024-08-21T10:09:23.033565581+09:00 [INFO][AMF][Ngap] Create a new NG connection for: 127.0.0.1:9487
2024-08-21T10:09:23.035146524+09:00 [INFO][AMF][Ngap][ran_addr:127.0.0.1:9487] Handle NGSetupRequest
2024-08-21T10:09:23.035240031+09:00 [INFO][AMF][Ngap][ran_addr:127.0.0.1:9487] Send NG-Setup response
2024-08-21T10:09:23.055287043+09:00 [INFO][AMF][Ngap][ran_addr:127.0.0.1:9487] Handle InitialUEMessage
2024-08-21T10:09:23.055369380+09:00 [INFO][AMF][Ngap][amf_ue_ngap_id:RU:1,AU:1(3GPP)][ran_addr:127.0.0.1:9487] New RanUe [RanUeNgapID:1][AmfUeNgapID:1]
2024-08-21T10:09:23.055450549+09:00 [INFO][AMF][Ngap][ran_addr:127.0.0.1:9487] 5GSMobileIdentity ["SUCI":"suci-0-208-93-0-0-0-00007487", err: <nil>]
2024-08-21T10:09:23.056785363+09:00 [INFO][AMF][CTX] New AmfUe [supi:][guti:20893cafe0000000001]
2024-08-21T10:09:23.056898047+09:00 [INFO][AMF][Gmm] Handle event[Gmm Message], transition from [Deregistered] to [Deregistered]
2024-08-21T10:09:23.056930796+09:00 [INFO][AMF][Gmm][amf_ue_ngap_id:RU:1,AU:1(3GPP)][supi:SUPI:] Handle Registration Request
2024-08-21T10:09:23.056961268+09:00 [INFO][AMF][Gmm][amf_ue_ngap_id:RU:1,AU:1(3GPP)][supi:SUPI:] RegistrationType: Initial Registration
2024-08-21T10:09:23.056990967+09:00 [INFO][AMF][Gmm][amf_ue_ngap_id:RU:1,AU:1(3GPP)][supi:SUPI:] MobileIdentity5GS: SUCI[suci-0-208-93-0-0-0-00007487]
2024-08-21T10:09:23.057043125+09:00 [INFO][AMF][Gmm] Handle event[Start Authentication], transition from [Deregistered] to [Authentication]
2024-08-21T10:09:23.057081486+09:00 [INFO][AMF][Gmm][amf_ue_ngap_id:RU:1,AU:1(3GPP)][supi:SUPI:] Authentication procedure
2024-08-21T10:09:23.058236725+09:00 [INFO][NRF][DISC] Handle NFDiscoveryRequest
2024-08-21T10:09:23.059914794+09:00 [INFO][NRF][GIN] | 200 |       127.0.0.1 | GET     | /nnrf-disc/v1/nf-instances?requester-nf-type=AMF&target-nf-type=AUSF |
2024-08-21T10:09:23.061390184+09:00 [INFO][AUSF][UeAuth] HandleUeAuthPostRequest
2024-08-21T10:09:23.061432664+09:00 [INFO][AUSF][UeAuth] Serving network authorized
2024-08-21T10:09:23.062310692+09:00 [INFO][NRF][DISC] Handle NFDiscoveryRequest
2024-08-21T10:09:23.065274230+09:00 [INFO][NRF][GIN] | 200 |       127.0.0.1 | GET     | /nnrf-disc/v1/nf-instances?requester-nf-type=AUSF&service-names=nudm-ueau&target-nf-type=UDM |
2024-08-21T10:09:23.066826371+09:00 [INFO][UDM][UEAU] Handle GenerateAuthDataRequest
2024-08-21T10:09:23.066911450+09:00 [INFO][UDM][Suci] suciPart: [suci 0 208 93 0 0 0 00007487]
2024-08-21T10:09:23.066936944+09:00 [INFO][UDM][Suci] scheme 0
2024-08-21T10:09:23.066958811+09:00 [INFO][UDM][Suci] SUPI type is IMSI
http://127.0.0.10:8000
2024-08-21T10:09:23.067693759+09:00 [INFO][NRF][DISC] Handle NFDiscoveryRequest
2024-08-21T10:09:23.069593774+09:00 [INFO][NRF][GIN] | 200 |       127.0.0.1 | GET     | /nnrf-disc/v1/nf-instances?requester-nf-type=UDM&target-nf-type=UDR |
2024-08-21T10:09:23.070852983+09:00 [INFO][UDR][DataRepo] Handle QueryAuthSubsData
2024-08-21T10:09:23.071531764+09:00 [INFO][UDR][GIN] | 200 |       127.0.0.1 | GET     | /nudr-dr/v1/subscription-data/imsi-2089300007487/authentication-data/authentication-subscription |
2024-08-21T10:09:23.072322764+09:00 [INFO][UDR][DataRepo] Handle ModifyAuthentication
2024-08-21T10:09:23.074748461+09:00 [INFO][UDR][GIN] | 204 |       127.0.0.1 | PATCH   | /nudr-dr/v1/subscription-data/imsi-2089300007487/authentication-data/authentication-subscription |
2024-08-21T10:09:23.075244561+09:00 [INFO][UDM][GIN] | 200 |       127.0.0.1 | POST    | /nudm-ueau/v1/suci-0-208-93-0-0-0-00007487/security-information/generate-auth-data |
2024-08-21T10:09:23.075525593+09:00 [INFO][AUSF][UeAuth] Add SuciSupiPair (suci-0-208-93-0-0-0-00007487, imsi-2089300007487) to map.
2024-08-21T10:09:23.075574553+09:00 [INFO][AUSF][UeAuth] Use 5G AKA auth method
2024-08-21T10:09:23.075604314+09:00 [INFO][AUSF][5gAka] XresStar = 3266616133333739663761646164326133643565323738363635323165643138
2024-08-21T10:09:23.075715925+09:00 [INFO][AUSF][GIN] | 201 |       127.0.0.1 | POST    | /nausf-auth/v1/ue-authentications |
2024-08-21T10:09:23.076052846+09:00 [INFO][AMF][Gmm][amf_ue_ngap_id:RU:1,AU:1(3GPP)][supi:SUPI:] Send Authentication Request
2024-08-21T10:09:23.076228857+09:00 [INFO][AMF][Ngap][amf_ue_ngap_id:RU:1,AU:1(3GPP)][ran_addr:127.0.0.1:9487] Send Downlink Nas Transport
2024-08-21T10:09:23.077138048+09:00 [INFO][AMF][Gmm][amf_ue_ngap_id:RU:1,AU:1(3GPP)][supi:SUPI:] Start T3560 timer
[]--- FAIL: TestReSynchronization (0.05s)
panic: runtime error: slice bounds out of range [:6] with capacity 0 [recovered]
        panic: runtime error: slice bounds out of range [:6] with capacity 0

goroutine 363 [running]:
testing.tRunner.func1.2({0x19b2220, 0xc000622b40})
        /usr/local/go/src/testing/testing.go:1545 +0x238
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1548 +0x397
panic({0x19b2220?, 0xc000622b40?})
        /usr/local/go/src/runtime/panic.go:914 +0x21f
github.com/free5gc/util/milenage.milenageF1({0xc000ec1b00, 0x10, 0x0?}, {0xc000ec1b20, 0x10, 0x20}, {0xc000ec1a30, 0x10, 0x20?}, {0x3296fe0, ...}, ...)
        /home/nishi/go/pkg/mod/github.com/free5gc/util@v1.0.6/milenage/milenage.go:43 +0x40f
github.com/free5gc/util/milenage.F1(...)
        /home/nishi/go/pkg/mod/github.com/free5gc/util@v1.0.6/milenage/milenage.go:509
test_test.TestReSynchronization(0xc000eab380?)
        /home/nishi/free5gc-deb/free5gc/test/registration_test.go:2614 +0xdee
testing.tRunner(0xc000eab380, 0x1bc0bb8)
        /usr/local/go/src/testing/testing.go:1595 +0xff
created by testing.(*T).Run in goroutine 1
        /usr/local/go/src/testing/testing.go:1648 +0x3ad
exit status 2
FAIL    test    1.988s
2024-08-21T10:09:26.136884060+09:00 [INFO][UPF][Main] Shutdown UPF ...
2024-08-21T10:09:26.137044784+09:00 [INFO][UPF][PFCP][LAddr:10.200.200.101:8805] Stopping pfcp server
2024-08-21T10:09:26.137177112+09:00 [ERRO][UPF][PFCP][LAddr:10.200.200.101:8805] read udp4 10.200.200.101:8805: use of closed network connection
2024-08-21T10:09:26.137244144+09:00 [INFO][UPF][PFCP][LAddr:10.200.200.101:8805] pfcp reciver stopped
2024-08-21T10:09:26.137322937+09:00 [INFO][UPF][PFCP][LAddr:10.200.200.101:8805] pfcp server stopped
2024-08-21T10:09:26.179273565+09:00 [INFO][UPF][Perio] recv event[TYPE_SERVER_CLOSE][{eType:4 lSeid:0 urrid:0 period:0}]
2024-08-21T10:09:26.179320921+09:00 [INFO][UPF][Perio] perio server stopped
2024-08-21T10:09:26.179363706+09:00 [INFO][UPF][Main] Terminating UPF...
2024-08-21T10:09:26.179391084+09:00 [INFO][UPF][Main] UPF terminated
2024-08-21T10:09:26.179414223+09:00 [INFO][UPF][Main] UPF exited
$
```

--- a/test/registration_test.go
+++ b/test/registration_test.go
@@ -1,11 +1,10 @@
 package test_test
 
 import (
-	"bytes"
 	"context"
+	cryptorand "crypto/rand"
 
 	"encoding/base64"
-	"encoding/binary"
 	"encoding/hex"
 	"fmt"
 	"net"
@@ -2574,52 +2573,30 @@ func TestReSynchronization(t *testing.T) {
 	K, OPC := make([]byte, 16), make([]byte, 16)
 	K, _ = hex.DecodeString(ue.AuthenticationSubs.PermanentKey.PermanentKeyValue)
 	OPC, _ = hex.DecodeString(ue.AuthenticationSubs.Opc.OpcValue)
-	SQN := make([]byte, 6)
-	AK := make([]byte, 6)
-
 	rand := nasPdu.AuthenticationRequest.GetRANDValue()
-	milenage.F2345(OPC, K, rand[:], nil, nil, nil, AK, nil)
-	autn := nasPdu.AuthenticationRequest.GetAUTN()
-	SQNxorAK := autn[:6]
-	for i := 0; i < 6; i++ {
-		SQN[i] = AK[i] ^ SQNxorAK[i]
-	}
-	const SqnMAx int64 = 0x7FFFFFFFFFF
-	const SqnMs int64 = 0
-	const IND int64 = 32
-	var newSqnMsString string
-	SQNBuffer := make([]byte, 8)
-	copy(SQNBuffer[2:], SQN)
-	r := bytes.NewReader(SQNBuffer)
-	var retrieveSqn int64
-	if err := binary.Read(r, binary.BigEndian, &retrieveSqn); err != nil {
-		fmt.Println("err", err)
-		return
-	}
 
-	delita := retrieveSqn - SqnMs
-	if delita < 0x7FFFFFFFFFF {
-		newSqnMsString = "000000000000"
-	}
+	// Based on TS 33.105, clause 5.1.1.3. The SQN_ms is a SQN value managed by the mobile station (or UE).
+	// Whenever the UE finds that SQN_ms is not in synced with SQN sent by the AMF, it start the
+	// re-synchronization process, sending AUTS (Authentication Token for Synchronization) to the
+	// AMF. AUTS is concatenation of concealed SQN_ms and MAC_S.
+	// AUTS = SQN_ms [^ AK*] | MAC-S
 
-	newSqnMs, _ := hex.DecodeString(newSqnMsString)
-	MAC_A, MAC_S := make([]byte, 8), make([]byte, 8)
-	CK, IK := make([]byte, 16), make([]byte, 16)
-	RES := make([]byte, 8)
-	AK, AKstar := make([]byte, 6), make([]byte, 6)
-	AMF, _ := hex.DecodeString("0000")
-	milenage.F1(OPC, K, rand[:], newSqnMs, AMF, MAC_A, MAC_S)
-	milenage.F2345(OPC, K, rand[:], RES, CK, IK, AK, AKstar)
+	// To test out that core network synchronize the SQN properly. We generated a random SQN here.
+	var randomSqnMs [6]byte
+	_, err = cryptorand.Read(randomSqnMs[:])
+	assert.Nil(t, err)
 
+	// Build AUTS
+	AK, MAC_S := make([]byte, 6), make([]byte, 8)
+	AMF, _ := hex.DecodeString("0000") // Re-Sync AMF
+	milenage.F1(OPC, K, rand[:], randomSqnMs[:], AMF, nil, MAC_S)
+	milenage.F2345(OPC, K, rand[:], nil, nil, nil, nil, AK)
 	SQNmsxorAK := make([]byte, 6)
-	for i := 0; i < len(SQN); i++ {
-		SQNxorAK[i] = SQN[i] ^ AK[i]
+	for i := 0; i < 6; i++ {
+		SQNmsxorAK[i] = randomSqnMs[i] ^ AK[i]
 	}
-	ColSQNmsxorAK := make([]byte, 6)
-	for i := 0; i < len(SQN); i++ {
-		ColSQNmsxorAK[i] = SQNmsxorAK[i] ^ AKstar[i]
-	}
-	AUTS := append(ColSQNmsxorAK, MAC_S...)
+	AUTS := append(SQNmsxorAK, MAC_S...)
+
 	// compute SQN by AUTN, K, AK
 	// suppose
 	// send NAS Authentication Rejcet
@@ -2645,10 +2622,12 @@ func TestReSynchronization(t *testing.T) {
 		"Received wrong GMM message. Expected Authentication Request.")
 	rand = nasPdu.AuthenticationRequest.GetRANDValue()
 
+	// After re-synchronization, check if the SQN is updated
 	milenage.F2345(OPC, K, rand[:], nil, nil, nil, AK, nil)
-	autn = nasPdu.AuthenticationRequest.GetAUTN()
-	SQNxorAK = autn[:6]
+	autn := nasPdu.AuthenticationRequest.GetAUTN()
+	SQNxorAK := autn[:6]
 
+	SQN := make([]byte, 6)
 	for i := 0; i < 6; i++ {
 		SQN[i] = AK[i] ^ SQNxorAK[i]
 	}

